### PR TITLE
[FLINK-16708][jdbc] When a JDBC connection has been closed, the retry…

### PR DESCRIPTION
**Pull request by shangwen**
## What is the purpose of the change

* This pull request adds a reconnection mechanism to JDBCUpsertOutputFormat, so that the flush method can retry successfully when the connection is closed *

## Brief change log

  - *Add reconnection mechanism in flush method*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by shangwen** [comment]
Hi @JingsongLi , could you please review this PR for me?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java [hunk 1]
```diff
@@ -166,6 +166,22 @@ public synchronized void flush() throws Exception {
...
+				try {
+					isValid = connection.isValid(1);
+				} catch (SQLException ex) {
+					LOG.warn("Check connection valid failed.", ex);
+				}
+
+				try {
+					if (!isValid) {
+						LOG.warn("This connection is not valid, trying to reconnect.");
+						establishConnection();
```
Comment (Tartarus0zm) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java:179]: Is it possible to put the reconnect operation in the exception handling of isValid?
if reconnect failed,then throw exception and retrying may be meaningless?

File: flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java [hunk 2]
```diff
@@ -199,7 +215,9 @@ public synchronized void close() throws IOException {
 		}
 
 		try {
-			jdbcWriter.close();
+			if (jdbcWriter != null) {
```
Comment (Tartarus0zm) [inline flink-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCUpsertOutputFormat.java:218]: In theory, jdbcWriter should not be null,

**Comment by rionmonster** [comment]
I see that this PR has been a bit stale - I've been running into this issue quite a bit recently and would be happy to either create a new pull request or make change to this one (depending on the level of drift, which might be significant). I've commented on the related JIRA ticket.

Thanks,

Rion

**Comment by MartijnVisser** [comment]
@shangwen @rionmonster We've now moved the code from the JDBC connector to its own dedicated repository. It would be great if one of you could open up a new PR for this ticket over there.

**Pull request closed without merging**